caas/kubernetes/provider/resources/inteface.go

```diff
@@ -0,0 +1,41 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package resources
+
+import (
+	"context"
+
+	"k8s.io/client-go/kubernetes"
+)
+
+//go:generate go run github.com/golang/mock/mockgen -package mocks -destination mocks/resources_mock.go github.com/juju/juju/caas/kubernetes/provider/resources Resource,Applier
+
+const (
+	// JujuFieldManager marks the resource changes were made by Juju.
+	JujuFieldManager = "juju"
+)
+
+// Resource defines methods for manipulating a k8s resource.
+type Resource interface {
```

Comment:
My preference is for a service oriented approach where there's entities which are used as args to various service methods (could be varidac for create, delete etc). Entities encapsulate state but not interactions with the cluster.
So there'd be a Resource interface with Clone(), String().
There'd be a Service interface with Apply(), Get(), Delete() - these take one or more entity resources as args.
Then there'd be a DeploymentService, a StatefulsetService, a DaemonsetService etc. These service instances hold a k8s client reference.

k8s itself takes this approach where there's a client that is akin to the service instance and the domain entities are the Deployment or Secret or whatever.

Reply:
I guess the Applier provides a different level of abstraction / operation grouping which is an alternative approach to take. So maybe my apprehension is misplaced.

Reply:
Lets chat about this tomorrow.

Reply:
As discussed, we keep the current approach for now.